KAFKA-15673: Adding client metrics resource types (KIP-714)

apoorvmittal10 opened the pull request:
This PR add resources to store and handle client metrics needed for KIP-714 as defined [here](https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability#KIP714:Clientmetricsandobservability-Clientmetricsconfiguration) - [jira](https://issues.apache.org/jira/browse/KAFKA-15673)

Changes include:
- Adding CLIENT_METRICS to resource type
- Corresponding DYNAMIC client configurations in resources.
- Changes to support dynamic loading of configuration on changes.
- Changes to support API calls to fetch data stored against the new resource.
- Test cases for the changes

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

apoorvmittal10 committed:
KAFKA-15673: Adding client metrics resource types with configuration handling

apoorvmittal10 commented on core/src/main/scala/kafka/server/ClientMetricsManager.scala:
```diff
@@ -0,0 +1,38 @@
+
+import java.util.Properties
+
+object ClientMetricsManager {
```
@hachikuji @mjsax Need feedback on new classes in broker to support client metrics on broker side, should they be in java or scala works?

cc: @AndrewJSchofield

apoorvmittal10 replied:
Moved to java.

apoorvmittal10 commented on core/src/main/scala/kafka/server/ClientMetricsManager.scala:
```diff
@@ -0,0 +1,38 @@
+object ClientMetricsManager {
+  private val Instance = new ClientMetricsManager
+
+  def getInstance: ClientMetricsManager = Instance
```
@hachikuji @mjsax @AndrewJSchofield Inherited some classes from earlier implementation of KIP and not sure if I should properly define trait for the client telemetry classes in broker.

apoorvmittal10 replied:
Moved to java

apoorvmittal10 commented on core/src/main/scala/kafka/metrics/clientmetrics/ClientMetricsConfig.scala:
```diff
@@ -0,0 +1,137 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.metrics.clientmetrics
```
@hachikuji @mjsax @AndrewJSchofield Do you think we should either move the classes to parent `metrics` package itself or something else. I didn't change the package created by @sarat-kakarla.

AndrewJSchofield replied:
I would move to the parent `metrics` package.

apoorvmittal10 replied:
I have moved it to parent `metrics` package.

AndrewJSchofield commented on core/src/main/scala/kafka/metrics/clientmetrics/ClientMetricsConfig.scala:
```diff
@@ -0,0 +1,137 @@
+    val PushIntervalMs = "interval.ms"
+    val ClientMatchPattern = "match"
+
+    val DEFAULT_PUSH_INTERVAL = 30 * 1000 // 5 minutes
```
I would call it `DEFAULT_INTERVAL_MS` so it's clear it's milliseconds. Also, the config is `interval.ms`.

apoorvmittal10 replied:
Done.

AndrewJSchofield commented on core/src/main/scala/kafka/metrics/clientmetrics/ClientMetricsConfig.scala:
```diff
@@ -0,0 +1,137 @@
+      // Make sure that push interval is between 100ms and 1 hour.
+      if (properties.containsKey(PushIntervalMs)) {
+        val pushIntervalMs = Integer.parseInt(properties.getProperty(PushIntervalMs))
+        if (pushIntervalMs < 100 || pushIntervalMs > 3600000)
```
There have to be constants for these min and max values that you can use.

apoorvmittal10 replied:
Done.

AndrewJSchofield commented on core/src/main/scala/kafka/metrics/clientmetrics/ClientMetricsConfig.scala:
```diff
@@ -0,0 +1,137 @@
+      if (properties.containsKey(PushIntervalMs)) {
+        val pushIntervalMs = Integer.parseInt(properties.getProperty(PushIntervalMs))
+        if (pushIntervalMs < 100 || pushIntervalMs > 3600000)
+          throw new InvalidRequestException(s"Invalid parameter ${PushIntervalMs}")
```
This exception would not mention which parameter was invalid. It would say something like `Invalid parameter 100`. Probably ought to improve it.

apoorvmittal10 replied:
Changed the message to capture values, please check.

AndrewJSchofield commented on core/src/main/scala/kafka/metrics/clientmetrics/ClientMetricsConfig.scala:
```diff
@@ -0,0 +1,137 @@
+
+    def names = configDef.names
+
+    def validate(subscriptionId :String, properties :Properties): Unit = {
```
Tiny formatting improvement. `subscriptionId: String, properties: Properties)` please.

apoorvmittal10 replied:
Done.

AndrewJSchofield commented on core/src/main/scala/kafka/metrics/clientmetrics/ClientMetricsMetadata.scala:
```diff
@@ -0,0 +1,130 @@
+   *     occurrence of the character '='
+   *  2. '*' is considered as invalid client match pattern
+   * @param patterns List of client matching pattern strings
+   * @return
```
`@return` what?

apoorvmittal10 replied:
Sorry, done.

AndrewJSchofield commented on core/src/main/scala/kafka/metrics/clientmetrics/ClientMetricsMetadata.scala:
```diff
@@ -0,0 +1,130 @@
+
+  private def validRegExPattern(inputPattern :String): Boolean = {
+    try {
+      Pattern.compile(inputPattern)
```
The KIP says specifically that the regex is compiled with Google RE2/J. This is generally what is used broker-side for regex for reasons of safety.

apoorvmittal10 replied:
I forgot to write this thing on the PR but I do not see re2/j being used in Kafka yet. I can see an old open ticket.

https://issues.apache.org/jira/browse/KAFKA-7276

apoorvmittal10 commented:
Build fails because of flaky tests

```
9 tests have failed
There are 0 new tests failing, 9 existing failing and 152 skipped.
```
```
Existing failures - 9
Build / JDK 21 and Scala 2.13 / testFollowerCompleteDelayedFetchesOnReplication(String).quorum=kraft – integration.kafka.server.FetchFromFollowerIntegrationTest
1s
Build / JDK 21 and Scala 2.13 / shouldHaveSamePositionBoundActiveAndStandBy – org.apache.kafka.streams.integration.ConsistencyVectorIntegrationTest
5s
Build / JDK 21 and Scala 2.13 / shouldAddNamedTopologyToRunningApplicationWithSingleInitialNamedTopology() – org.apache.kafka.streams.integration.NamedTopologyIntegrationTest
1m 2s
Build / JDK 21 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest
2m 0s
Build / JDK 8 and Scala 2.12 / testCreateTokenForOtherUserFails(String).quorum=kraft – kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest
3s
Build / JDK 8 and Scala 2.12 / testListenerConnectionRateLimitWhenActualRateAboveLimit() – kafka.network.ConnectionQuotasTest
19s
Build / JDK 11 and Scala 2.13 / shouldHaveSamePositionBoundActiveAndStandBy – org.apache.kafka.streams.integration.ConsistencyVectorIntegrationTest
16s
Build / JDK 11 and Scala 2.13 / shouldManualUpgradeFromNonVersionedNonTimestampedToVersioned – org.apache.kafka.streams.integration.VersionedKeyValueStoreIntegrationTest
1m 5s
Build / JDK 11 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest

AndrewJSchofield commented on core/src/main/scala/kafka/server/ConfigHelper.scala:
```diff
@@ -57,13 +58,15 @@ class ConfigHelper(metadataCache: MetadataCache, config: KafkaConfig, configRepo
         case ConfigResource.Type.TOPIC =>
           authHelper.authorize(request.context, DESCRIBE_CONFIGS, TOPIC, resource.resourceName)
+        case ConfigResource.Type.CLIENT_METRICS =>
+          authHelper.authorize(request.context, DESCRIBE_CONFIGS, CLIENT_METRICS, resource.resourceName)
```
The KIP says that the ACL checked is operation `DESCRIBE_CONFIGS` and resource `CLUSTER`.

apoorvmittal10 replied:
Completely missed it, thanks for that. I have removed declaring new Resource Type for client_metrics and corresponding checks. Have changed test cases to authorize on CLUSTER resource for CLIENT_METRICS config source.

AndrewJSchofield commented on core/src/main/scala/kafka/server/ControllerApis.scala:
```diff
@@ -483,6 +483,12 @@ class ControllerApis(
           new ApiError(TOPIC_AUTHORIZATION_FAILED)
         }
+      case ConfigResource.Type.CLIENT_METRICS =>
+        if (authHelper.authorize(requestContext, ALTER_CONFIGS, CLIENT_METRICS, resource.name)) {
```
The KIP says that the ACL check is operation `ALTER` and resource `CLUSTER`. I think that perhaps the KIP should actually say `ALTER_CONFIGS` but the resource type should remain `CLUSTER`. OK?

apoorvmittal10 replied:
Corrected it.

AndrewJSchofield commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2754,6 +2754,8 @@ class KafkaApis(val requestChannel: RequestChannel,
         case ConfigResource.Type.TOPIC =>
           authHelper.authorize(originalRequest.context, ALTER_CONFIGS, TOPIC, resource.name)
+        case ConfigResource.Type.CLIENT_METRICS =>
+          authHelper.authorize(originalRequest.context, ALTER_CONFIGS, CLIENT_METRICS, resource.name)
```
See earlier comment. There are no ACLs for `CLIENT_METRICS` resource type and checking is performed against the `CLUSTER` resource.

apoorvmittal10 replied:
Completely missed it, thanks for that. I have removed declaring new Resource Type for client_metrics and corresponding checks. Have changed test cases to authorize on CLUSTER resource for CLIENT_METRICS config source.

AndrewJSchofield commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2916,6 +2918,8 @@ class KafkaApis(val requestChannel: RequestChannel,
         case ConfigResource.Type.TOPIC =>
           authHelper.authorize(originalRequest.context, ALTER_CONFIGS, TOPIC, resource.name)
+        case ConfigResource.Type.CLIENT_METRICS =>
+          authHelper.authorize(originalRequest.context, ALTER_CONFIGS, CLIENT_METRICS, resource.name)
```
And `CLUSTER` again here too.

apoorvmittal10 replied:
Done, thanks.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/common/resource/ResourceType.java:
```diff
@@ -68,7 +68,12 @@ public enum ResourceType {
+    /**
+     *  Client metrics
+     */
+    CLIENT_METRICS((byte) 8);
```
The KIP does not define this constant. Checking is against the `CLUSTER` resource type.

apoorvmittal10 replied:
Missed this thing, thanks. Removed and corrected tests.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented on core/src/main/scala/kafka/metrics/ClientMetricsConfig.scala:
```diff
@@ -0,0 +1,141 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.metrics
```
@AndrewJSchofield @hachikuji @sarat-kakarla @mjsax Does it make more sense to move this class (as java) in `org.apache.kafka.common.config.` or `org.apache.kafka.common.config.internals` rather than in `core`. I see all configuration classes are defined in `common` itself.

apoorvmittal10 replied:
Atleast the config def to be moved?

apoorvmittal10 replied:
Moved to java.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2619,6 +2619,9 @@ private ConfigEntry.ConfigSource configSource(DescribeConfigsResponse.ConfigSour
             case DYNAMIC_BROKER_LOGGER_CONFIG:
                 configSource = ConfigEntry.ConfigSource.DYNAMIC_BROKER_LOGGER_CONFIG;
                 break;
+            case DYNAMIC_CLIENT_METRICS_CONFIG:
```
I think I would make this match the `TOPIC_CONFIG` so that this value is actually `CLIENT_METRICS_CONFIG`. They're all dynamic, just like topic configs.

apoorvmittal10 replied:
Make sense, done.

AndrewJSchofield commented on core/src/main/scala/kafka/metrics/ClientMetricsMetadata.scala:
```diff
@@ -0,0 +1,130 @@
+                   clientId: String,
+                   softwareName: String,
+                   softwareVersion: String,
+                   clientHostAddress: String,
```
I suggest to make this `clientSourceAddress` and `clientSourcePort` so the names all agree with the match selector names in the KIP.

apoorvmittal10 replied:
Dne, also changed other variables.

AndrewJSchofield commented on core/src/main/scala/kafka/server/ZkConfigManager.scala:
```diff
@@ -38,7 +38,8 @@ object ConfigType {
   val Broker = "brokers"
   val Ip = "ips"
-  val all = Seq(Topic, Client, User, Broker, Ip)
+  val ClientMetrics = "client-metrics"
```
Given that we don't support client metrics on ZK, I am surprised to see code changes in ZkConfigManager.

apoorvmittal10 replied:
This is a bit strange I know but ConfigType is defined under this class.

philipnee commented on core/src/main/scala/kafka/metrics/ClientMetricsMetadata.scala:
```diff
@@ -0,0 +1,130 @@
+  }
+  def getClientId: Option[String] = attributesMap.get(CLIENT_ID)
+
+  def isMatched(patterns: Map[String, String]) : Boolean = {
```
An alternative way to avoid null and if else :
```
def isMatched(patterns: Option[Map[String, String]]): Boolean = {
  patterns match {
    case Some(p) if p.nonEmpty => matchPatterns(p)
    case _ => true
  }
}
```

apoorvmittal10 replied:
Thanks for the suggestion, done.

philipnee commented on core/src/main/scala/kafka/metrics/ClientMetricsMetadata.scala:
```diff
@@ -0,0 +1,130 @@
+    try {
+      matchingPatterns.foreach {
+      case (k, v) =>
+        val attribute = attributesMap.getOrElse(k, null)
```
invalidConfiguraitonException is not needed, if you use pattern matching, you can do:

```
attributesMap.get(k) match {
case Some(attr) if v.r.anchored.findFirstIn(attr).isDefined => 
          true
case _ => 
          info(s"Unable to find the matching client...")
          false
```

philipnee replied:
Scratch the one above, maybe something like this?

```
matchingPatterns.find {
 case (k,v) =>
   attributeMap.get(k) match {
       case Some(attr) if v.r.anchored.findFirstMatchIn(attr).isDefined => false
       case _ => true
   }
}.isEmpty

apoorvmittal10 replied:
I have simplified the code, see if it makes more sense now or still you want to go with your suggestion.

philipnee commented on core/src/test/scala/kafka/metrics/ClientMetricsTestUtils.scala:
```diff
@@ -0,0 +1,39 @@
+import java.util.Properties
+
+object ClientMetricsTestUtils {
+  val defaultPushInterval = 30 * 1000 // 30 seconds
```
most of the time-vars here are suffix with Ms, I wonder if we want to do it here as well...

apoorvmittal10 replied:
Done.

philipnee commented on core/src/main/scala/kafka/metrics/ClientMetricsConfig.scala:
```diff
@@ -0,0 +1,141 @@
+ * should push the metrics to the broker.
+ * <li> "match" is a comma separated list of client match patterns, in case if there is no matching
+ * pattern specified then broker considers that as all match which means the associated metrics
+ * applies to all the clients. Ex: "client_software_name = Java, client_software_version = 11.1.*"
```
The example suggests * is supported.  In this case, all 11.1 subversion will be returned right? It would be useful to mention that in the doc.

Would it work for metrics? Can I do "org.apache.kafka.client.producer.*" ?

apoorvmittal10 replied:
I have added the example for match and added prefix match details for metrics. Done.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 committed:
Fixing refactor issue

philipnee commented on core/src/main/scala/kafka/metrics/ClientMetricsMetadata.scala:
```diff
@@ -0,0 +1,119 @@
+    val clientSoftwareName = if (ctx.clientInformation != null) ctx.clientInformation.softwareName() else ""
+    val clientSoftwareVersion = if (ctx.clientInformation != null) ctx.clientInformation.softwareVersion() else ""
+    instance.init(clientInstanceId, ctx.clientId(), clientSoftwareName, clientSoftwareVersion,
+                  ctx.clientAddress.getHostAddress, ctx.clientAddress.getHostAddress) // TODO: Fix Port
```
ctx.clientAddress.getHostAddress, ctx.clientAddress.getHostAddress -> You are passing in two hostAddress, this perhaps is the TODO item?

apoorvmittal10 replied:
Yes, I will address this in next PR.

philipnee commented on core/src/main/scala/kafka/server/ClientMetricsManager.scala:
```diff
@@ -0,0 +1,38 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
Extra line.

apoorvmittal10 replied:
Done.

philipnee commented on core/src/main/scala/kafka/metrics/ClientMetricsConfig.scala:
```diff
@@ -0,0 +1,146 @@
+ */
+object ClientMetricsConfig {
+
+  class SubscriptionInfo(subscriptionId: String,
```
I would recommend using scala `case class` here.  It provides you with the basic utils such as getter.

philipnee replied:
I think this is how you can write this:

```
case class SubscriptionInfo(subscriptionId: String,
                              subscribedMetrics: List[String],
                              pushIntervalMs: Int,
                              matchingPatternsList: List[String]) { 
    def clientMatchingPatterns: Map[String, String] = 
      ClientMetricsMetadata.parseMatchingPatterns(matchingPatternsList)
  }
```

Does it accept an empty subscriptionId?

apoorvmittal10 replied:
SubscriptionId should not be empty. Good suggestion, I ll keep the comment as open for now as I have asked @hachikuji that should we place these new classes in Java itself as I see new code is written there. In case we go with scala then I ll update the code with suggestion.

apoorvmittal10 replied:
Moved to java implementation

philipnee commented on core/src/test/scala/kafka/metrics/ClientMetricsTestUtils.scala:
```diff
@@ -0,0 +1,39 @@
+import java.util.Properties
+
+object ClientMetricsTestUtils {
+  val defaultPushIntervalMs: Int = 30 * 1000 // 30 seconds
```
Do you want to make these constants camel starting with a capital letter like what you did in `ClientMetrics`? In general in scala, `Constant names should be in upper camel case` per scala documentation.

apoorvmittal10 replied:
Done, my bad.

philipnee commented on core/src/main/scala/kafka/metrics/ClientMetricsMetadata.scala:
```diff
@@ -0,0 +1,119 @@
+}
+
+class ClientMetricsMetadata {
+  var attributesMap: mutable.Map[String, String] = scala.collection.mutable.Map[String, String]()
```
Do you need the init? If you don't, you can just use the primary class constructor:
```
class ClientMetricsMetadata(clientInstanceId: String,
                            clientId: String,
                            clientSoftwareName: String,
                            clientSoftwareVersion: String,
                            clientSourceAddress: String,
                            clientSourcePort: String) {
  private val attributesMap: mutable.Map[String, String] = mutable.Map(
    CLIENT_INSTANCE_ID -> clientInstanceId,
    CLIENT_ID -> clientId,
    CLIENT_SOFTWARE_NAME -> clientSoftwareName,
    CLIENT_SOFTWARE_VERSION -> clientSoftwareVersion,
    CLIENT_SOURCE_ADDRESS -> clientSourceAddress,
    CLIENT_SOURCE_PORT -> clientSourcePort
  )
```

such that the apply method above can just be 

```
def apply(...): ClientMetricsMetadata = {
    new ClientMetricsMetadata(...)
  }
```

apoorvmittal10 replied:
Valid, I ll take it up once decided to continue this class in scala or move to respective java package.

apoorvmittal10 replied:
Moved to java implementation

philipnee commented:
Hi @apoorvmittal10 Thanks for taking my suggestions.  I left a few more, mostly about the aesthetic aspect as scala is a pretty idiomatic language...

Terribly sorry that I don't have a lot of knowledge around the server side, so I can't provide advice on detail implementation - but I think your PR is clean and I can almost fully understand the intention.

philipnee commented on core/src/main/scala/kafka/metrics/ClientMetricsMetadata.scala:
```diff
@@ -0,0 +1,119 @@
+    val patternsMap = mutable.Map[String, String]()
+    if (patterns != null) {
+      patterns.foreach(x => {
+        val nameValuePair = x.split("=", 2).map(x => x.trim)
```
a cool thing you can do is pattern matching, you can match `x.split("=", 2).map(x => x.trim)` with a pair `(k, v)` and direct the non-matching case `case _`

This is what I meant
```
pattern.split("=", 2).map(_.trim) match {
 case (k, v) if isValidParam(k) && validRegexPattern(v) => patternMap + (k -> v)
 case _ => throw new InvalidConfigurationException()
}
```

apoorvmittal10 replied:
Again, thanks. I ll take it up once decided to continue this class in scala or move to respective java package.

apoorvmittal10 committed:
Fixing variable names

apoorvmittal10 commented:
> Hi @apoorvmittal10 Thanks for taking my suggestions. I left a few more, mostly about the aesthetic aspect as scala is a pretty idiomatic language...
> 
> Terribly sorry that I don't have a lot of knowledge around the server side, so I can't provide advice on detail implementation - but I think your PR is clean and I can almost fully understand the intention.

Thanks for the review. It's helpful and using scala for the first time, sorry about issues. It ll get better from next PRs. I haven't yet addressed your suggestions in new classes as I have asked @hachikuji if it makes more sense to write then in Java, as I see new classes for `core` are now written in Java itself.
cc: @junrao

apoorvmittal10 committed:
Correcting default time

junrao commented:
@apoorvmittal10 : Yes, we are gradually moving the codebase to java. Ideally, all new classes should be written in java.

apoorvmittal10 commented:
> @apoorvmittal10 : Yes, we are gradually moving the codebase to java. Ideally, all new classes should be written in java.

Thanks @junrao, I ll update the PR by tomorrow with new classes in Java.

apoorvmittal10 committed:
Movinf new scala classes to java

apoorvmittal10 commented:
> > @apoorvmittal10 : Yes, we are gradually moving the codebase to java. Ideally, all new classes should be written in java.
> 
> Thanks @junrao, I ll update the PR by tomorrow with new classes in Java.

@junrao @hachikuji I have moved new classes to java, Can I please get a review on the PR.

junrao commented on core/src/main/scala/kafka/server/ConfigHelper.scala:
```diff
@@ -129,6 +130,25 @@ class ConfigHelper(metadataCache: MetadataCache, config: KafkaConfig, configRepo
+
+          case ConfigResource.Type.CLIENT_METRICS =>
+            val entityName = resource.resourceName
+            if (resource.resourceName == null || resource.resourceName.isEmpty) {
```
Could we use `entityName`?

apoorvmittal10 replied:
I have renamed it to `subscriptionName` and used that for checks, thanks for pointing out.

junrao commented on core/src/main/java/kafka/metrics/ClientMetricsConfigs.java:
```diff
@@ -0,0 +1,181 @@
+        if (properties.containsKey(PUSH_INTERVAL_MS)) {
+            int pushIntervalMs = Integer.parseInt(properties.getProperty(PUSH_INTERVAL_MS));
+            if (pushIntervalMs < MIN_INTERVAL_MS || pushIntervalMs > MAX_INTERVAL_MS) {
+                throw new InvalidRequestException("Invalid value for interval.ms, interval must be between 100 and 3600000 (1 hour)");
```
Could we include `pushIntervalMs` in the error message?

apoorvmittal10 replied:
Added in the message.

junrao commented on core/src/main/java/kafka/metrics/ClientMetricsConfigs.java:
```diff
@@ -0,0 +1,181 @@
+        if (patterns != null) {
+            patterns.forEach(pattern -> {
+                String[] nameValuePair = pattern.split("=");
+                if (nameValuePair.length == 2 && isValidParam(nameValuePair[0].trim()) &&
```
Could we avoid calling `nameValuePair[0].trim()` twice?

apoorvmittal10 replied:
Done.

junrao commented on core/src/test/java/kafka/metrics/ClientMetricsTestUtils.java:
```diff
@@ -0,0 +1,41 @@
+public class ClientMetricsTestUtils {
+
+    public static final String DEFAULT_METRICS =
+        "org.apache.kafka.client.producer.partition.queue.,org.apache.kafka.client.producer.partition.latency";
```
extra dot at the end of org.apache.kafka.client.producer.partition.queue.?

apoorvmittal10 replied:
That's I have written knowingly to match something defined in KIP already as example. The subscribed metrics shall be a prefix match.

https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability#KIP714:Clientmetricsandobservability-Metricssubscription

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -0,0 +1,39 @@
+/**
+ * Handles client telemetry metrics requests/responses, subscriptions and instance information.
+ */
+public class ClientMetricsManager {
```
Do we need to close ClientMetricsManager on shutdown?

apoorvmittal10 replied:
We might need that and I have added `Closeable` in the class. I will implement that along other code for ClientMetricsManager.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -106,7 +106,8 @@ class KafkaApis(val requestChannel: RequestChannel,
                 val tokenManager: DelegationTokenManager,
-                val apiVersionManager: ApiVersionManager
+                val apiVersionManager: ApiVersionManager,
+                val clientMetricsManager: ClientMetricsManager
```
Will `clientMetricsManager` be used in future PRs?

apoorvmittal10 replied:
Yes, it will be used. It will be the major class to handle KIP-714 telemetry APIs calls.

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -567,7 +572,8 @@ class KafkaServer(
                                                            ConfigType.Broker -> new BrokerConfigHandler(config, quotaManagers),
-                                                           ConfigType.Ip -> new IpConfigHandler(socketServer.connectionQuotas))
+                                                           ConfigType.Ip -> new IpConfigHandler(socketServer.connectionQuotas),
+                                                           ConfigType.ClientMetrics -> new ClientMetricsConfigHandler(clientMetricsManager))
```
Do we need this since the KIP says the feature is only supported in KRaft mode?

apoorvmittal10 replied:
Thanks for pointing out, yeah it seems this code is just for ZK config handlers. I have removed the code from here.

junrao commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -4089,13 +4119,19 @@ public void testIncrementalAlterConfigs()  throws Exception {
                     new ConfigEntry("compression.type", "gzip"),
                     AlterConfigOp.OpType.APPEND);
 
+            AlterConfigOp alterConfigOp3 = new AlterConfigOp(
```
This doesn't seem to be a valid config for CLIENT_METRICS?

apoorvmittal10 replied:
I have corrected the config as per CLIENT_METRICS.

junrao commented on core/src/test/scala/unit/kafka/server/ControllerConfigurationValidatorTest.scala:
```diff
@@ -101,4 +102,55 @@ class ControllerConfigurationValidatorTest {
+    config.put(ClientMetricsConfigs.PUSH_INTERVAL_MS, "2000")
+    config.put(ClientMetricsConfigs.SUBSCRIPTION_METRICS, "org.apache.kafka.client.producer.partition.queue.,org.apache.kafka.client.producer.partition.latency")
+    config.put(ClientMetricsConfigs.CLIENT_MATCH_PATTERN, "client_instance_id=b69cc35a-7a54-4790-aa69-cc2bd4ee4538,client_id=1" +
+      ",client_software_name=apache-kafka-java,client_software_version=2.8.0-SNAPSHOT,client_source_address=127.0.0.1," +
```
We use `apache-kafka-java` here, but `java` in `ClientMetricsTestUtils`. What's the standard name?

apoorvmittal10 replied:
`apache-kafka-java` is the standard name, I have corrected `ClientMetricsTestUtils`.

junrao commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -477,6 +479,128 @@ class KafkaApisTest {
+      new Action(operation, new ResourcePattern(resourceType, Resource.CLUSTER_NAME, PatternType.LITERAL),
+        1, true, true)
+    )
+    // Verify that authorize is only called once
```
The verification happens later in line 593.

apoorvmittal10 replied:
Corrected.

junrao commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -2619,6 +2619,9 @@ private ConfigEntry.ConfigSource configSource(DescribeConfigsResponse.ConfigSour
             case DYNAMIC_BROKER_LOGGER_CONFIG:
                 configSource = ConfigEntry.ConfigSource.DYNAMIC_BROKER_LOGGER_CONFIG;
                 break;
+            case CLIENT_METRICS_CONFIG:
```
Hmm, this method seems very specific to get/create topic. Do we really need to add client metric here?

apoorvmittal10 replied:
You are right, this should not be needed. I have removed it, I am not sure why other config handlers are required here when the method is just used by topics code. May be previously it was used elsewhere as well.

junrao replied:
Some of the configs could be set at both topic and broker levels. So, there is a hierarchy.

apoorvmittal10 replied:
Sounds good, thanks for letting me know.

junrao commented on core/src/test/scala/unit/kafka/server/ControllerConfigurationValidatorTest.scala:
```diff
@@ -101,4 +102,55 @@ class ControllerConfigurationValidatorTest {
+  }
+
+  @Test
+  def testInvalidSubscriptionIdClientMetricsConfig(): Unit = {
```
It's testing invalid subscription name instead of invalid subscription id.

apoorvmittal10 replied:
Corrected the method name.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Left a few more comments.

apoorvmittal10 committed:
Addressing review comments

apoorvmittal10 commented:
> @apoorvmittal10 : Thanks for the updated PR. Left a few more comments.

Thanks for the feedback @junrao. I have addressed the comments.

junrao commented on clients/src/main/java/org/apache/kafka/clients/admin/ConfigEntry.java:
```diff
@@ -222,6 +222,7 @@ public enum ConfigSource {
         DYNAMIC_BROKER_LOGGER_CONFIG,   // dynamic broker logger config that is configured for a specific broker
         DYNAMIC_BROKER_CONFIG,          // dynamic broker config that is configured for a specific broker
         DYNAMIC_DEFAULT_BROKER_CONFIG,  // dynamic broker config that is configured as default for all brokers in the cluster
+        DYNAMIC_CLIENT_METRICS_CONFIG,  // dynamic client metrics subscription config that is configured for all clients
```
It seems weird to put `CLIENT_METRICS` here. The rest of the entries all seem to be related to configs that could be set in different levels like topic, broker, broker default, etc. Other entity types like user, client, etc are not listed here. Ditto for the change in `DescribeConfigsResponse` and `ConfigResource`.

apoorvmittal10 replied:
I might be wrong but shouldn't this be needed to make `kafka-config.sh` work to create `subscription` for client_metrics. To support the `kafka-config.sh` utility we would require change in `ConfigEntry` and `ConfigSource`. I have this PR that uses same: https://github.com/apache/kafka/pull/14632

junrao replied:
@apoorvmittal10 : Thanks for the comment. I mis-understood the logic here. The create/describe config RPC only applies to topic/broker/BrokerLoggerConfigType and we are adding CLIENT_METRICS here. quota/Ip level configs are handled by different types of PRC. So, the logic you have here is fine.

junrao commented on core/src/main/java/kafka/server/ClientMetricsManager.java:
```diff
@@ -0,0 +1,46 @@
+    private static final Logger log = LoggerFactory.getLogger(ClientMetricsManager.class);
+    private static final ClientMetricsManager INSTANCE = new ClientMetricsManager();
+
+    public static ClientMetricsManager getInstance() {
```
We don't use getters. So this could just be `instance`. Ditto for `getDefaultProperties` in `ClientMetricsTestUtils`.

apoorvmittal10 replied:
Done for both `getInstance` and `getDefaultProperties`.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Left a couple of more comments.

apoorvmittal10 committed:
Removing get from methods

junrao commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -543,7 +547,8 @@ class KafkaServer(
           tokenManager = tokenManager,
-          apiVersionManager = apiVersionManager)
+          apiVersionManager = apiVersionManager,
+          clientMetricsManager = clientMetricsManager)
```
Since client metric is not supported in ZK mode, should we make clientMetricsManager an Option and pass in None here?

apoorvmittal10 replied:
I agree, it makes sense to not use KafkaServer as that's specifically in ZK mode. I have made the param optional and pushed the change.

junrao reviewed (commented):
@apoorvmittal10 : Thanks for the updated PR. Just one more comment.

apoorvmittal10 committed:
Making metrics manager optional

apoorvmittal10 commented:
> @apoorvmittal10 : Thanks for the updated PR. Just one more comment.

Thanks a lot for the feedback, it was helpful. I have addressed the comments.

apoorvmittal10 committed:
Merge remote-tracking branch 'upstream/trunk' into kip-714-ak-broker

apoorvmittal10 commented:
@junrao I can find these 7 unrelated tests from the PR are failing.

```
Build / JDK 17 and Scala 2.13 / testTransactionAfterTransactionIdExpiresButProducerIdRemains(String).quorum=kraft – kafka.api.ProducerIdExpirationTest
8s
Build / JDK 8 and Scala 2.12 / testBumpTransactionalEpoch(String).quorum=kraft – kafka.api.TransactionsTest
1m 20s
Build / JDK 11 and Scala 2.13 / testReplicateSourceDefault() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationBaseTest
2m 15s
Build / JDK 11 and Scala 2.13 / testReplicateSourceDefault() – org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest
1m 51s
Build / JDK 11 and Scala 2.13 / testDeleteCmdNonExistingGroup(String).quorum=kraft – kafka.admin.DeleteConsumerGroupsTest
11s
Build / JDK 11 and Scala 2.13 / testTimeouts() – org.apache.kafka.controller.QuorumControllerTest
<1s
Build / JDK 11 and Scala 2.13 / testHighWaterMarkAfterPartitionReassignment(String).quorum=kraft – org.apache.kafka.tools.reassign.ReassignPartitionsIntegrationTest

junrao reviewed (approved):
@apoorvmittal10 : Thanks for the updated PR. LGTM.

For the unrelated failed tests, could we file jiras to track new ones? Thanks.

apoorvmittal10 commented:
> @apoorvmittal10 : Thanks for the updated PR. LGTM.
> 
> For the unrelated failed tests, could we file jiras to track new ones? Thanks.

Thanks @junrao. I have added this jira for tracking: https://issues.apache.org/jira/browse/KAFKA-15785

junrao commented:
@apoorvmittal10 : Thanks for the filing the jira. It's better to file a separate jira for each new failed test.

apoorvmittal10 commented:
> @apoorvmittal10 : Thanks for filing the jira. It's better to file a separate jira for each new failed test.

Sure, done.
